### PR TITLE
prov/psm3: Reserve API values for psm3 provider

### PR DESCRIPTION
**Pull request by acgoldma**
Signed-off-by: Goldman, Adam <adam.goldman@intel.com>

### CODE REVIEW

**Comment by jsquyres** [comment]
Now that OPA and OPA2 have moved to https://www.cornelisnetworks.com/, is PSM3 related to that?  Does it run on the same hardware?  What's the relationship between OPA, OPA2, and OPA3?

**Comment by mwheinz** [comment]
> Now that OPA and OPA2 have moved to https://www.cornelisnetworks.com/, is PSM3 related to that? Does it run on the same hardware? What's the relationship between OPA, OPA2, and OPA3?

The original PSM or "PSM 1" is effectively dead, it was for old QLogic IB hardware is no longer supported by either Intel or Cornelis. PSM2 is used by OPA hardware and software and is currently supported by Cornelis. I cannot speak for "PSM3" this is something new.

**Comment by tschuett** [comment]
https://github.com/intel/eth-psm3-fi

**Comment by acgoldma** [comment]
> > Now that OPA and OPA2 have moved to https://www.cornelisnetworks.com/, is PSM3 related to that? Does it run on the same hardware? What's the relationship between OPA, OPA2, and OPA3?
> 
> The original PSM or "PSM 1" is effectively dead, it was for old QLogic IB hardware that is no longer supported by either Intel or Cornelis. PSM2 is used by OPA hardware and software and is currently supported by Cornelis. I cannot speak for "PSM3" this is something new.

PSM3 is the supported by Intel and runs over Verbs with enhanced support for Intel's Ethernet Network Adapter E810 Series. 
An out-of-tree provider can be found: https://github.com/intel/eth-psm3-fi

**Comment by jsquyres** [comment]
Might I suggest that "PSM3" is quite a confusing choice of name for the library:

* It's completely unrelated to the hardware that ran PSM1 and PSM2
* It's (now) from a completely different vendor than PSM1 and PSM2 (even if PSM1 is "dead")
* The name PSM3 implies a relationship to PSM1 and PSM2 that doesn't exist.

Please consider changing the name.

**Comment by acgoldma** [comment]
> Might I suggest that "PSM3" is quite a confusing choice of name for the library:
> 
> * It's completely unrelated to the hardware that ran PSM1 and PSM2
> * It's (now) from a completely different vendor than PSM1 and PSM2 (even if PSM1 is "dead")
> * The name PSM3 implies a relationship to PSM1 and PSM2 that doesn't exist.
> 
> Please consider changing the name.

PSM3 can run on most verbs devices that support UD traffic.

PSM is an acronym for Performance Scaled Messaging which is a protocol still owned by Intel.  While Intel has licensed some parts of the technology to others, Intel still retains all the IP rights and ownership. 
PSM3 is the 3rd version of this protocol that now targets the verbs layer. There were actually quite a lot of differences in the previous HW generations (SDR, DDR, QDR, OPA100) over the 20 year life of the PSM protocol and it’s concepts.  Introducing standard RDMA below PSM protocol was a logical next step and retains all the key attributes, algorithms, concepts and approaches from PSM2.

The PSM3 name was carefully chosen to reflect the legacy and maturity of the protocol and implementation.  Initial customer feedback regarding the name has been positive.

File: src/common.c [hunk 1]
```diff
@@ -462,6 +469,24 @@ static int ofi_str_to_psmx2(const char *str, void **addr, size_t *len)
...
+{
+	int ret;
+
+	*len = 2 * sizeof(uint64_t);
+	*addr = calloc(1, *len);
+	if (!(*addr))
+		return -FI_ENOMEM;
+
+	ret = sscanf(str, "%*[^:]://%" SCNx64 ":%" SCNx64,
+		     (uint64_t *) *addr, (uint64_t *) *addr + 1);
```
Comment (shefty) [inline src/common.c:482]: This is fine, and is what several other providers do.  But does the address break into components that would be easier for a user to handle?  For example, see FI_SOCKADDR_IN, FI_SOCKADDR_IB, and FI_ADDR_IB_UD as an example for how the addressing can be broken into more decipherable pieces.

**Comment by jsquyres** [comment]
I'm not your customer and I have no stake in this.  I'm just a random guy who's been working in this ecosystem for a while, including supporting a range of customers -- those who are highly skilled and deeply familiar with the entire breadth and depth of the stack, those who don't know/care about any of this and are just trying to run numerical simulations at scale, and lots of customers in between those two extremes.

My point is that it's confusing to have a name "FOO3" that implies a continuity with "FOO2" (and presumably "FOO1"), but really isn't.  This happened with PSM and PSM2, as well -- many users (and non-Intel developers) assumed "Oh, PSM2?  That's version 2 of the PSM library... we should be using that because 2>1, right?"

This required an explanation to users.  Every.  Single.  Time.

You're now doing the same thing again: PSM3 is _not_ an upgrade to PSM or PSM2, even though the industry-common software library naming convention implies that it is.  Worse, the PSM ecosystem now spans two vendors, and PSM3 is now a wholly different ecosystem (vendor hardware and software) that has -- from the sysadmin/network admin perspective -- nothing to do with two other product lines that happen to have incredibly similar software support library names.

> The PSM3 name was carefully chosen to reflect the legacy and maturity of the protocol and implementation. Initial customer feedback regarding the name has been positive.

The distinction between PSM, PSM2, and PSM3 may be inherently obvious to those of you who work deeply in that stack / ecosystem.  But that distinction is readily lost upon the rest of the world.

As I said, I'm not your customer and you can totally ignore me.  But I'm posting this because I really, really do not want to have several more years of having to explain to *your customers* the difference between PSM, PSM2, and PSM3.

**Comment by shefty** [comment]
Intel CI is a known hang that's being debugged and can be ignored here.

File: src/common.c (hunk 1 above)
Comment (acgoldma) [reply src/common.c:482]: Unfortunately, the 128 bit address is a combination items like in psm2 which are mostly opaque to the user.

**Comment by shefty** [comment]
Ignoring any personal preference and stepping into a neutral maintainer role here, IMO it's ultimately up to the provider (and their marketing department) as to what they want to call their provider, within reason.  For instance, I would reject offensive names or names trademarked by another company.  Since @acgoldma indicates that the PSM name is still owned by Intel, I don't have an objection as a maintainer.

**Comment by ToddRimmer** [comment]
As background and to answer some of @jsquyres questions

As always, if customer questions come up, please direct them to Intel or this discussion list for clarification.  We will try to make  this clear in the pdf's which will be available for free on the Intel download center where we also make Intel supported binaries available for customers for free.

I agree there can sometimes be confusion around the PSM name, is it a protocol, a software API or a hardware specific driver.  It has provided variations of all 3 in the past. 

I believe the convention in the industry is that changes to the major revision number can indicate major new features and removal of support for older hardware.  To this extent, PSM1 supported some old IB hardware, PSM1 also had only a direct API.  PSM2 dropped support for that old IB hardware, added some API features (notably larger matching fields to allow for growing rank counts) and added support for new HW.  PSM2 was available as a direct API and an OFI provider of the same name.  As PSM2 progressed the API received minor revisions and the latest OPA PSM2 API rev was 2.2.  The binary revs progressed fast as new tunings and bug fixes were added.  I agree there were mistakes in the PSM1 to PSM2 library progression and some clunky approaches to retain API forward compatibility (infnipath compat library).  

PSM3 is now a continuation of the same software concepts, algorithms, and higher level aspects of the protocol, but has dropped the direct API (only offering OFI, hence no compat library issues) and has dropped support for OPA (and IB) hardware and their wire formats in favor of standard RoCE.

We have successfully validated PSM3 against middleware which previously used OFI psm2 provider with no changes to the middleware.  As is always the case, there may be opportunities for middleware to tune specifically for psm3 and the new HW it supports and middleware which bypasses or incompletely uses the OFI API may encounter limitations.  We'll assist to workaround or correct such limitations as they are uncovered, for example we recently did some work for openmpi in PSM3 to workaround the auth_key gaps in it's use of OFI API.  For now PSM3 also checks for some env variables currently exported by openmpi 4.x which had the same information.  Intel has engaged the openmpi community and PXMI maintainer on that detail.

**Comment by jsquyres** [comment]
@ToddRimmer I hear what you're saying.  But the fact that you had to write a 5-paragraph explanation pretty much guarantees that this level of nuance will be lost on the general population, and those of us in the consumer communities (e.g., MPI) will be left to mop up the confusion (again).

Saying "direct all questions to us/Intel" doesn't even help, because the consumer libraries (e.g., MPI) are always blamed first (and then maybe we'll figured out its an OFI problem, and then maybe we'll figure out its a PSM/PSM2/PSM3 problem).  Additionally, there's now 2 vendors involved -- when should we direct questions to Cornelius?  When should we direct questions to Intel?  And why the heck do we -- the customers and the consumer community -- have to figure this out?  If your transport library had a unique name, this would take at least some of the variables out of this calculus (e.g., it would avoid multiple rounds of back-n-forth with the customer -- "I have a PSM network." / "Which PSM?" / "How do I figure that out?" / "...").

In short: I realize that the distinctions that you made are clear, logical, and have a progression.  It's still a pretty complicated story, and is now even more complicated because there's multiple vendors involved.  99% of the rest of the world won't make these distinctions.

**Comment by shefty** [comment]
Picked into #6529
